pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,249 @@
+	reserveRadius = rs.radius(is)
+	return
+}
+
+// tier returns which tier a value falls into
+func (rs *reserveState) tier(x *big.Int) tier {
+	if x.Cmp(rs.Inner) < 0 || rs.Inner.Cmp(big.NewInt(0)) == 0 {
+		return none
+	}
+	if x.Cmp(rs.Outer) < 0 {
+		return half
+	}
+	return full
+}
+
+// radius returns the reserve radius of a batch given the depth (radius of responsibility)
+// based on the tier it falls in
+func (rs *reserveState) radius(t tier) uint8 {
+	switch t {
+	case none:
+		return swarm.MaxPO
+	case half:
+		return rs.Depth
+	default: // full
+		return rs.Depth - 1
+	}
+}
+
+// setLimits sets the tier 1 value limit, if new item is the minimum so far (or the very first batch)
+func (rs *reserveState) setLimits(val *big.Int, was tier) (is tier) {
+	if was != none {
+		return was
+	}
+	if rs.Inner.Cmp(big.NewInt(0)) == 0 || rs.Inner.Cmp(val) > 0 {
+		is = half
+		if rs.Outer.Cmp(rs.Inner) == 0 {
+			rs.Outer.Set(val)
+			is = full
+		}
+		rs.Inner.Set(val)
+	}
+	return is
+}
+
+// update manages what chunks of which batch are allocated to the reserve
+func (s *store) update(b *postage.Batch, oldDepth uint8, oldValue *big.Int) error {
+	newValue := b.Value
+	newDepth := b.Depth
+	capacityChange, reserveRadius := s.rs.change(oldValue, newValue, oldDepth, newDepth)
+	s.rs.Capacity += capacityChange
+	if err := s.unreserve(b, reserveRadius); err != nil {
+		return err
+	}
+	return s.evictOuter(b)
+}
+
+// evictOuter is responsible for keeping capacity positive by unreserving lowest priority batches
+func (s *store) evictOuter(last *postage.Batch) error {
+	// if capacity is positive nothing to evict
+	if s.rs.Capacity >= 0 {
+		return nil
+	}
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		batchID := valueKeyToID(key)
+		b := last
+		if !bytes.Equal(b.ID, batchID) {
+			var err error
+			b, err = s.Get(batchID)
+			if err != nil {
+				return true, fmt.Errorf("release get %x %v: %w", batchID, b, err)
+			}
+		}
+		//  FIXME: this is needed only because  the statestore iterator does not allow seek, only prefix
+		//  so we need to page through all the batches until outer limit is reached
+		if b.Value.Cmp(s.rs.Outer) < 0 {
+			return false, nil
+		}
+		// stop iteration  only if we consumed all batches of the same value as the one that put capacity above zero
+		if s.rs.Capacity >= 0 && s.rs.Outer.Cmp(b.Value) != 0 {
+			return true, nil
+		}
+		//
+		// unreserve outer PO of the lowest priority batch  until capacity is back to positive
+		s.rs.Capacity += exp2(b.Depth - s.rs.Depth - 1)
+		s.rs.Outer.Set(b.Value)
+		return false, s.unreserve(b, s.rs.Depth)
+	})
+	if err != nil {
+		return err
+	}
+	// add 1 to outer limit value so we dont hit on the same batch next time we iterate
+	s.rs.Outer.Add(s.rs.Outer, big1)
+	// if we consumed all batches, ie. we unreserved all chunks on the outer = depth PO
+	//  then its time to  increase depth
+	if s.rs.Capacity < 0 {
```

Comment:
i don't understand how is it that `Capacity` can be negative, or what that means

Reply:
capacity -n means we need to make space for n chunks